tools/ontology-builder/src/compute_descendent_mappings.py

```diff
@@ -0,0 +1,279 @@
+#!/usr/bin/env python
+"""
+# Descendant Mappings for Tissues and Cell Types
+
+## Overview
+
+The ontology-aware tissue and cell type filters in the Single Cell Data Portal each require artifacts generated
+by this script.
+
+#### Descendant Mappings
+To facilitate in-filter, cross-panel restriction of filter values, a descendant hierarchy dictionary is required by
+the Single Cell Data Portal frontend. For example, if a user selects `hematopoietic system` in the tissue filter's
+`System` panel, the values in the tissue filter's `Organ` and `Tissue` panels must be restricted by `hematopoietic
+system`.
+
+This script generates a dictionary of descendants keyed by tissue or cell type ontology term ID. The dictionary
+is stored as a JSON file and copied to cellxgene-ontology-guide/ontology_assets directory. A versioned github release is
+ created to simplify referencing in the Single Cell Data Portal.
+
+The descendant mappings should be updated when:
+
+1. The ontology version is updated,
+2. A new tissue or cell type is added to the production corpus, or,
+3. The hand-curated systems, organs, cell classes or cell subclasses are updated.
+"""
+
+import json
+import os
+from typing import Any, Dict, List
+from urllib.request import urlopen
+
+import env
+from cellxgene_ontology_guide.ontology_parser import OntologyParser
+
+
+def load_prod_datasets() -> Any:
+    """
+    Request datasets the production corpus.
+    """
+    response = urlopen("https://api.cellxgene.cziscience.com/dp/v1/datasets/index").read().decode("utf-8")
+    return json.loads(response)
+
+
+def extract_cell_types(datasets: List[Dict[str, Any]]) -> List[str]:
+    """
+    List the set of cell type values for the given datasets.
+
+    :param datasets: a list of datasets from the production corpus.
+    :return: a list formated of cell type values
+    """
+    cell_types = set()
+    for dataset in datasets:
+        for cell_type in dataset["cell_type"]:
+            cell_types.add(cell_type["ontology_term_id"].replace("_", ":", False))
+    return list(cell_types)
+
+
+def extract_tissues(datasets: List[Dict[str, Any]]) -> List[str]:
+    """
+    List the set of tissue values for the given datasets.
+
+    :param datasets: a list of datasets from the production corpus.
+    :return: a list of formated tissue values with tags for tissue type.
+    """
+    tissues = set()
+    for dataset in datasets:
+        for tissue in dataset["tissue"]:
+            formatted_entity_name = tissue["ontology_term_id"].replace("_", ":", False)
+            tissue_type = tissue.get("tissue_type")
+            tissues.add(tag_tissue_type(formatted_entity_name, tissue_type))
+
+    return list(tissues)
+
+
+def tag_tissue_type(entity_name: str, tissue_type: str) -> str:
```

Comment:
nit: can we replace instances of entity_name with ontology_term_id or similar? its confusing because term IDs also have names, which are not the ID. But I believe we are referring to the term ID here and elsewhere when we say entity_name

Reply:
I'll change it to term_id unless multiple term_id types exists, then i will user ontology_term_id